### PR TITLE
Codecs, plugins, system channels, GLFW-fixes

### LINKED ISSUE #35: Plugins
I'm wondering once StandardMethodCodec is in place, if this library could use https://github.com/GoogleCloudPlatform/golang-samples to implement the Firebase plugins found at github.com/flutter/plugins ?

### PR DESCRIPTION
**Pull request by GeertJohan**
Apologies for the huge change-set. This PR contains a lot of changes, including some breaking changes.

Some breaking changes could not be properly avoided.

Summary of changes:
- Adds a plugin package with message codecs, method codecs, binary messenger interface, handler interfaces, and channels. Includes unit tests based on flutter's unit tests.
- Adds a flutter.Plugin interface which is now the way to add plugins to a go-flutter application.
- Reworked platform plugins (clipboard, textediting, window title) to work using the new Plugin system.
- Deprecated some options.
- Small typo fixes and improved logging.
- TODO notes on things that should be looked at.

**This PR is still WIP. There are still some regressions, I'm testing and changing the packages outlines.**

**I'm constantly re-basing this branch, please don't commit based on this branch, or contact me beforehand so we can work together without merging/conflict issues**

Fixes #35

### CODE REVIEW

File: embedder/build.go [hunk 1]
```diff
@@ -18,8 +18,3 @@ package embedder
```
Comment (Drakirus) [inline embedder/build.go]: CFLAGS should be `-I${SRCDIR}` and not `-I${SRCDIR}/library`

File: embedder/embedder_proxy_glfw.go [hunk 2]
```diff
@@ -9,32 +9,23 @@ static char *c_str(uint8_t *str){
...
 
 	"github.com/go-gl/glfw/v3.2/glfw"
 )
 
 // C proxies
 
-//export proxy_on_platform_message
-func proxy_on_platform_message(message *C.FlutterPlatformMessage, window unsafe.Pointer) C.bool {
-	if message.message != nil {
-		str := C.GoStringN(C.c_str(message.message), C.int(message.message_size))
```
Comment (Drakirus) [inline embedder/embedder_proxy_glfw.go]: `c_str` function can be removed

**Comment by GeertJohan** [comment]
I'm running into an issue with async handling of methodcalls. I've opened an issue to find out where it goes wrong as I couldn't find good documentation on it. flutter/flutter#29444

**Comment by GeertJohan** [comment]
Meanwhile this PR contains more changes:
- Limit CPU usage by using timeouts on glfw wait event, ~60fps
- Add thread-safety by locking the main handling goroutine to the thread it starts the flutter engine on.
- Added a simple tasker system that runs some tasks on the main goroutine (calls to glfw and flutter engine must occur from the same thread they started on).
- inlined the textModel into the textinputPlugin
- Added a callback on glfw to redraw the window when it was obscured by another window.
- More improvements in the codecs and plugin structure.

**Comment by GeertJohan** [comment]
I've also tagged most in-code TODO's with a github issue number. Those TODO's wont be done as part of this PR. I'd really like to get this merged because it's already way too big. Waiting for comments on flutter/flutter#29444 so I may resolve that issue, then this should be good to go.

**Comment by GeertJohan** [comment]
I have removed the async handling of platform messages for the time being. This may be added later when there's more clarity on how to properly handle them. This means that the package layout may change. API's should still be considered unstable/alpha for the time being.

**Review by pchampio — APPROVED** [review]
LGTM

**Pull request merged**